File: web-app/src/services/students.service.ts
/**
 * ============================================================================
 * STUDENTS SERVICE - API Client pour le module Scolarité (Élèves)
 * ============================================================================
 */

import { apiFetch } from "@/lib/api/client";
import { createEntityOffline, updateEntityOffline } from "@/lib/offline/offline-business.service";
import { networkDetectionService } from "@/lib/offline/network-detection.service";
import { LocalSearchService } from "@/lib/offline/local-search.service";

const BASE_URL = "/students";

function getTenantId(): string {
  if (typeof document === "undefined") return "";
  const match = document.cookie.match(/(?:(?:^|.*;\s*)x-tenant-id\s*\=\s*([^;]*).*$)|^.*$/);
  return match ? decodeURIComponent(match[1]) : "";
}

class StudentsService {
  /**
   * Récupère la liste des étudiants
   */
  async getAll(params?: Record<string, string>): Promise<any> {
    const qs = params ? new URLSearchParams(params).toString() : "";
    
    if (!networkDetectionService.isConnected()) {
      return LocalSearchService.search("students", { tenantId: getTenantId() });
    }

    try {
      return await apiFetch(`${BASE_URL}${qs ? `?${qs}` : ""}`);
    } catch (error) {
      return LocalSearchService.search("students", { tenantId: getTenantId() });
    }
  }

  /**
   * Admissions
   */
  async getAdmissions(params?: Record<string, string>): Promise<any> {
    const qs = params ? new URLSearchParams(params).toString() : "";
    return apiFetch(`${BASE_URL}/admissions${qs ? `?${qs}` : ""}`);
  }

  async getAdmissionById(id: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}`);
  }

  async createAdmission(data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions`, {
      method: "POST",
      body: data,
    });
  }

  async updateAdmission(id: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: data,
    });
  }

  async submitAdmission(id: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/submit`, {
      method: "POST",
    });
  }

  async decideAdmission(
    id: string,
    data: { decision: "ACCEPTED" | "REJECTED"; comment: string },
  ): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/decide`, {
      method: "POST",
      body: data,
    });
  }

  async convertAdmission(id: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/admissions/${encodeURIComponent(id)}/convert`,
      {
        method: "POST",
      },
    );
  }

  /**
   * Crée un nouvel étudiant
   */
  async create(data: any): Promise<any> {
    // Mode Offline-First complet pour les opérations CRUD basiques
    const tenantId = getTenantId();
    if (tenantId) {
      return createEntityOffline(tenantId, "STUDENT", data);
    }
    
    // Fallback si tenantId non trouvé
    return apiFetch(BASE_URL, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Met à jour un étudiant
   */
  async update(id: string, data: any): Promise<any> {
    const tenantId = getTenantId();
    if (tenantId) {
      return updateEntityOffline(tenantId, "STUDENT", id, data);
    }
    
    // Fallback si tenantId non trouvé
    return apiFetch(`${BASE_URL}/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: data,
    });
  }

  /**
   * Récupère les inscriptions
   */
  async getEnrollments(params?: Record<string, string>): Promise<any> {
    const qs = params ? new URLSearchParams(params).toString() : "";
    return apiFetch(`${BASE_URL}/enrollments${qs ? `?${qs}` : ""}`);
  }

  /**
   * Crée une inscription / admission
   */
  async createEnrollment(data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/enrollments`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Pré-inscription
   */
  async preRegister(data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/pre-register`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Ajouter des tuteurs
   */
  async addGuardians(
    studentId: string,
    data: { guardians: any[] },
  ): Promise<any> {
    return apiFetch(`${BASE_URL}/${encodeURIComponent(studentId)}/guardians`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Finaliser l'inscription (enroll)
   */
  async enrollStudent(studentId: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/${encodeURIComponent(studentId)}/enroll`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Statistiques des cartes d'identité
   */
  async getIdCardStats(academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/id-cards/stats?academicYearId=${academicYearId}`,
    );
  }

  /**
   * Génération par lot de cartes d'identité
   */
  async generateBulkIdCards(data: {
    academicYearId: string;
    schoolLevelId: string;
  }): Promise<any> {
    return apiFetch(`${BASE_URL}/id-cards/generate-bulk`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Statistiques des matricules
   */
  async getMatriculeStats(): Promise<any> {
    return apiFetch(`${BASE_URL}/identifiers/stats`);
  }

  /**
   * Génération par lot de matricules
   */
  async generateBulkMatricules(data: {
    academicYearId: string;
    schoolLevelId: string;
    status?: string;
  }): Promise<any> {
    return apiFetch(`${BASE_URL}/identifiers/generate-bulk?countryCode=BJ`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Générer un matricule individuel
   */
  async generateMatricule(studentId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/identifiers/${encodeURIComponent(studentId)}/generate?countryCode=BJ`,
      {
        method: "POST",
      },
    );
  }

  /**
   * Rechercher par matricule
   */
  async searchByMatricule(matricule: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/identifiers/search/${encodeURIComponent(matricule)}`,
    );
  }

  /**
   * Statistiques Générales
   */
  async getStatistics(
    academicYearId: string,
    schoolLevelId: string,
  ): Promise<any> {
    const params = new URLSearchParams({ academicYearId, schoolLevelId });
    return apiFetch(`${BASE_URL}/statistics?${params}`);
  }

  /**
   * KPIs ORION
   */
  async getOrionKpis(academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/orion/kpis?academicYearId=${encodeURIComponent(academicYearId)}`,
    );
  }

  /**
   * Alertes ORION
   */
  async getOrionAlerts(academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/orion/alerts?academicYearId=${encodeURIComponent(academicYearId)}`,
    );
  }

  /**
   * Révoquer une carte d'identité
   */
  async revokeIdCard(id: string, reason: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/id-cards/card/${encodeURIComponent(id)}/revoke`,
      {
        method: "PUT",
        body: { reason },
      },
    );
  }

  /**
   * Dossier et historique
   */
  async getDossier(id: string, academicYearId?: string): Promise<any> {
    const url = `/api/students/${id}/dossier${academicYearId ? `?academicYearId=${academicYearId}` : ""}`;
    return apiFetch(url);
  }

  async getHistory(id: string): Promise<any> {
    return apiFetch(`${BASE_URL}/${encodeURIComponent(id)}/history`);
  }

  async getVerificationQR(id: string, academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/${encodeURIComponent(id)}/verification-qr?academicYearId=${encodeURIComponent(academicYearId)}`,
    );
  }

  async regenerateVerificationQR(
    id: string,
    academicYearId: string,
  ): Promise<any> {
    return apiFetch(
      `${BASE_URL}/${encodeURIComponent(id)}/verification-token/regenerate`,
      {
        method: "POST",
        body: { academicYearId },
      },
    );
  }

  /**
   * Télécharge le dossier académique PDF
   */
  async downloadAcademicDossier(
    id: string,
    academicYearId: string,
  ): Promise<void> {
    const res = await fetch(
      `/api/students/${encodeURIComponent(id)}/academic-dossier?academicYearId=${encodeURIComponent(academicYearId)}`,
      {
        headers: {
          // Need to add auth headers manually if not using apiClient for blob
          "x-tenant-id": localStorage.getItem("tenantId") || "",
        },
      },
    );
    if (!res.ok) throw new Error("Erreur lors du téléchargement");
    const blob = await res.blob();
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `DOSSIER_${id}.pdf`;
    document.body.appendChild(a);
    a.click();
    window.URL.revokeObjectURL(url);
    a.remove();
  }

  /**
   * Export Educmaster Excel
   */
  async exportEducmasterExcel(
    academicYearId: string,
    schoolLevelId: string,
  ): Promise<void> {
    const res = await fetch(
      `/api/students/export/educmaster-excel?academicYearId=${encodeURIComponent(academicYearId)}&schoolLevelId=${encodeURIComponent(schoolLevelId)}`,
      {
        headers: {
          "x-tenant-id": localStorage.getItem("tenantId") || "",
        },
      },
    );
    if (!res.ok) {
      const err = await res
        .json()
        .catch(() => ({ message: "Erreur lors de l'export" }));
      throw new Error(err.message || "Erreur lors de l'export");
    }
    const blob = await res.blob();
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `EDUCMASTER_EXPORT_${schoolLevelId}_${academicYearId}.xlsx`;
    document.body.appendChild(a);
    a.click();
    window.URL.revokeObjectURL(url);
    a.remove();
  }

  /**
   * Upload la photo d'un étudiant
   * Note: FormData ne peut pas être stringifié, donc on ajuste method et headers.
   */
  async uploadPhoto(formData: FormData): Promise<any> {
    // Si FormData, apiFetch ou apiClient gère automatiquement le header multipart/form-data
    return apiFetch(`${BASE_URL}/upload-photo`, {
      method: "POST",
      body: formData,
      headers: {
        "Content-Type": undefined, // Let Axios set the correct boundary
      },
    });
  }
}

export const studentsService = new StudentsService();

File: apps/mobile-app/lib/data/services/students_service.dart
/// ============================================================================
/// STUDENTS SERVICE — Academia Hub Mobile
/// ============================================================================
///
/// Service CRUD pour le module Scolarité (Élèves).
/// Miroir de students.service.ts du web app.
///
/// Méthodes :
/// - getAll, create, update, delete (CRUD de base)
/// - getDossier, getHistory (Dossier & historique)
/// - getAdmissions, createAdmission, updateAdmission, submitAdmission,
///   decideAdmission, convertAdmission (Admissions)
/// - getEnrollments, createEnrollment, enrollStudent (Inscriptions)
/// - preRegister, addGuardians (Pré-inscription & tuteurs)
/// - getIdCardStats, generateBulkIdCards, revokeIdCard (Cartes d'identité)
/// - getMatriculeStats, generateBulkMatricules, generateMatricule,
///   searchByMatricule (Matricules)
/// - getStatistics, getOrionKpis, getOrionAlerts (Statistiques & ORION)
/// - uploadPhoto (Photo)
/// ============================================================================

import '../../core/crud/base_crud_service.dart';
import '../../core/network/api_config.dart';
import '../../core/network/api_result.dart';
import '../../core/network/api_client.dart';
import '../../core/offline/offline_service.dart';

class StudentsService extends BaseCrudService {
  StudentsService()
      : super(
          endpoint: ApiConfig.students,
          collection: 'students',
          entityType: SyncEntityType.student,
        );

  // ─── Admissions ──────────────────────────────────────────────────────────

  /// Récupère la liste des admissions.
  Future<ApiResult<List<Map<String, dynamic>>>> getAdmissions({
    Map<String, dynamic>? params,
  }) async {
    return apiGetWithFallback(
      ApiConfig.studentAdmissions,
      params: params,
      localFilters: {'type': 'admission'},
    );
  }

  /// Récupère une admission par ID.
  Future<ApiResult<Map<String, dynamic>>> getAdmissionById(String id) async {
    return getById(id);
  }

  /// Crée une nouvelle admission.
  Future<ApiResult<Map<String, dynamic>>> createAdmission(
    Map<String, dynamic> data,
  ) async {
    return apiPostWithOfflineFallback(
      ApiConfig.studentAdmissions,
      {...data, 'admissionStatus': 'PENDING'},
    );
  }

  /// Met à jour une admission.
  Future<ApiResult<Map<String, dynamic>>> updateAdmission(
    String id,
    Map<String, dynamic> data,
  ) async {
    return apiUpdateWithOfflineFallback(
      ApiConfig.studentAdmissionById(id),
      id,
      data,
    );
  }

  /// Soumet une admission pour validation.
  Future<ApiResult<Map<String, dynamic>>> submitAdmission(String id) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentAdmissionSubmit(id),
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Prend une décision sur une admission (ACCEPTED / REJECTED).
  Future<ApiResult<Map<String, dynamic>>> decideAdmission(
    String id, {
    required String decision,
    required String comment,
  }) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentAdmissionDecide(id),
        data: {'decision': decision, 'comment': comment},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Convertit une admission en inscription.
  Future<ApiResult<Map<String, dynamic>>> convertAdmission(String id) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentAdmissionConvert(id),
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  // ─── Inscriptions ────────────────────────────────────────────────────────

  /// Récupère la liste des inscriptions.
  Future<ApiResult<List<Map<String, dynamic>>>> getEnrollments({
    Map<String, dynamic>? params,
  }) async {
    return apiGetWithFallback(
      ApiConfig.studentEnrollments,
      params: params,
      localFilters: {'enrollmentStatus': 'ENROLLED'},
    );
  }

  /// Crée une inscription.
  Future<ApiResult<Map<String, dynamic>>> createEnrollment(
    Map<String, dynamic> data,
  ) async {
    return apiPostWithOfflineFallback(
      ApiConfig.studentEnrollments,
      {...data, 'type': 'enrollment'},
    );
  }

  /// Pré-inscription.
  Future<ApiResult<Map<String, dynamic>>> preRegister(
    Map<String, dynamic> data,
  ) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentPreRegister,
        data: data,
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Ajoute des tuteurs à un élève.
  Future<ApiResult<Map<String, dynamic>>> addGuardians(
    String studentId,
    List<Map<String, dynamic>> guardians,
  ) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentGuardians(studentId),
        data: {'guardians': guardians},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Finalise l'inscription d'un élève.
  Future<ApiResult<Map<String, dynamic>>> enrollStudent(
    String studentId,
    Map<String, dynamic> data,
  ) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentEnroll(studentId),
        data: data,
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  // ─── Cartes d'identité ──────────────────────────────────────────────────

  /// Statistiques des cartes d'identité.
  Future<ApiResult<Map<String, dynamic>>> getIdCardStats(
    String academicYearId,
  ) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentIdCardStats,
        queryParameters: {'academicYearId': academicYearId},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Génération par lot de cartes d'identité.
  Future<ApiResult<Map<String, dynamic>>> generateBulkIdCards({
    required String academicYearId,
    required String schoolLevelId,
  }) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentGenerateBulkIdCards,
        data: {
          'academicYearId': academicYearId,
          'schoolLevelId': schoolLevelId,
        },
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Révoque une carte d'identité.
  Future<ApiResult<Map<String, dynamic>>> revokeIdCard(
    String id,
    String reason,
  ) async {
    try {
      final response = await ApiClient.instance.put(
        ApiConfig.studentRevokeIdCard(id),
        data: {'reason': reason},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  // ─── Matricules ─────────────────────────────────────────────────────────

  /// Statistiques des matricules.
  Future<ApiResult<Map<String, dynamic>>> getMatriculeStats() async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentIdentifiersStats,
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Génération par lot de matricules.
  Future<ApiResult<Map<String, dynamic>>> generateBulkMatricules({
    required String academicYearId,
    required String schoolLevelId,
    String? status,
  }) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentGenerateBulkMatricules,
        data: {
          'academicYearId': academicYearId,
          'schoolLevelId': schoolLevelId,
          if (status != null) 'status': status,
        },
        queryParameters: {'countryCode': 'BJ'},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Génère un matricule individuel.
  Future<ApiResult<Map<String, dynamic>>> generateMatricule(
    String studentId,
  ) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentGenerateMatricule(studentId),
        queryParameters: {'countryCode': 'BJ'},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Recherche par matricule.
  Future<ApiResult<Map<String, dynamic>>> searchByMatricule(
    String matricule,
  ) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentSearchByMatricule(matricule),
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  // ─── Dossier & Historique ────────────────────────────────────────────────

  /// Récupère le dossier complet d'un élève.
  Future<ApiResult<Map<String, dynamic>>> getDossier(
    String id, {
    String? academicYearId,
  }) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentDossier(id),
        queryParameters: academicYearId != null
            ? {'academicYearId': academicYearId}
            : null,
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Récupère l'historique d'un élève.
  Future<ApiResult<List<Map<String, dynamic>>>> getHistory(String id) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentHistory(id),
      );
      return ApiSuccess(
        (response.data as List).map((e) => e as Map<String, dynamic>).toList(),
      );
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Récupère le QR code de vérification.
  Future<ApiResult<Map<String, dynamic>>> getVerificationQR(
    String id,
    String academicYearId,
  ) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentVerificationQR(id),
        queryParameters: {'academicYearId': academicYearId},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Régénère le token de vérification.
  Future<ApiResult<Map<String, dynamic>>> regenerateVerificationQR(
    String id,
    String academicYearId,
  ) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentRegenerateQR(id),
        data: {'academicYearId': academicYearId},
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  // ─── Statistiques ───────────────────────────────────────────────────────

  /// Statistiques générales des élèves.
  Future<ApiResult<Map<String, dynamic>>> getStatistics(
    String academicYearId,
    String schoolLevelId,
  ) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentStatistics(academicYearId, schoolLevelId),
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// KPIs ORION pour les élèves.
  Future<ApiResult<Map<String, dynamic>>> getOrionKpis(
    String academicYearId,
  ) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentOrionKpis(academicYearId),
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  /// Alertes ORION pour les élèves.
  Future<ApiResult<List<Map<String, dynamic>>>> getOrionAlerts(
    String academicYearId,
  ) async {
    try {
      final response = await ApiClient.instance.get(
        ApiConfig.studentOrionAlerts(academicYearId),
      );
      return ApiSuccess(
        (response.data as List).map((e) => e as Map<String, dynamic>).toList(),
      );
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }

  // ─── Photo ───────────────────────────────────────────────────────────────

  /// Upload la photo d'un élève.
  Future<ApiResult<Map<String, dynamic>>> uploadPhoto(
    String studentId,
    FormData formData,
  ) async {
    try {
      final response = await ApiClient.instance.post(
        ApiConfig.studentUploadPhoto(studentId),
        data: formData,
        options: Options(headers: {'Content-Type': 'multipart/form-data'}),
      );
      return ApiSuccess(response.data as Map<String, dynamic>);
    } catch (e) {
      return ApiFailure(ApiError.fromDioException(e));
    }
  }
}
